sympy/polys/rings.py

```diff
@@ -643,8 +584,83 @@ def _gens(self):
 
         return tuple(generators)
 
+    # to be overridden
+    def _gen_index(self, gen):
+        # get generator index from int or str
+        raise NotImplementedError("Must be implemented by subclass")
+
+    def add_gens(self, symbols):
+        """Add new generator(s) to the ring."""
+        raise NotImplementedError("Must be implemented by subclass")
+
+    def drop(self, *gens):
+        """Remove specified generator(s) from the ring."""
+        raise NotImplementedError("Must be implemented by subclass")
+
     def _ring_equality(self, other):
-        # Check equality of two polynomial rings.
+        # for ring attribute comparison
+        raise NotImplementedError("Must be implemented by subclass")
+
+    def _from_dict_ground(self, element, orig_domain=None):
+        # create polynomial from dict
+        raise NotImplementedError("Must be implemented by subclass")
+
+    def drop_to_ground(self, *gens):
+        """Remove generators and inject them into the ground domain."""
+        indices = set(map(self.index, gens))
+        symbols = [s for i, s in enumerate(self.symbols) if i not in indices]
+        gens_to_drop = [gen for i, gen in enumerate(self.gens) if i not in indices]
+
+        if not symbols:
+            return self
+        else:
+            return self.clone(symbols=symbols, domain=self.drop(*gens_to_drop))
+
+
+class PythonPolyRing(PolyRing):
+    """Multivariate distributed polynomial rings."""
+
+    def __new__(cls, symbols, domain, order=lex):
+        # validate inputs
+        symbols = tuple(_parse_symbols(symbols))
+        domain = DomainOpt.preprocess(domain)
+        order = OrderOpt.preprocess(order)
+
+        # initialize instance
+        obj = object.__new__(cls)
+        obj._init_instance(symbols, domain, order)
+        return obj
```

Comment:
I think it is better not to define `__new__` in the subclass. There can be a single `__new__` method in the superclass that preprocesses the arguments and dispatches to a class method like `._new()` for the appropriate subclass. If we have multiple subclasses then they should not duplicate things like `DomainOpt.preprocess(domain)`.